Add a prototype for asciidoc

sean- opened the pull request:
@arekinath , can you give this a quick once over w/ your tooling.  I had to go digging into the spec to see how multiple authors are supported (semicolon delimited list), but I don't know how the missing `:email:` attribute will impact things (or if that should also be a semicolon delimited list).

Feel free to merge if you approve.

sean- committed:
Add a prototype for asciidoc

jclulow commented:
I suspect we still want the YAML front matter block to be the same as it would be for a Markdown document?

sean- committed:
Revise the prototype asciidoc to match latest thinking.

Chase @tRent's change in 390ccb4f339172eda4671812989c4ef61f04ea43.

sean- commented:
@jclulow I looked at RFD 77 and didn't see it so I went with what was there and blessed.  Mixing and matching markups for the sake of Github rendering isn't something I'd be happy with.

[update] I looked at what @trentm did a few hours ago re: RFD77 and assumed that to be the canonical state of the world.  As our thinking evolves we can track updates and make the necessary adjustments.

jclulow commented:
It is, as far as I can tell, a content-agnostic set of YAML front matter block that we're expecting to put everywhere.  While Github _can_ render it, I believe the form originally came from Jekyll.

jclulow commented:
Note that it appears that [asciidoctor does have support for handling the front matter block](http://asciidoctor.org/docs/user-manual/#front-matter-added-for-static-site-generators) on at least some level.

trentm commented:
@arekinath and I discussed the usage of an AsciiDoc custom attribute (i.e. `:state: $state`) in private chat yesterday.  I wasn't aware at the time that AsciiDoc supported Jekyll front matter (thanks for the link @jclulow!). Having the same format for both .adoc and .md RFDs would be slightly easier for tools, but either would work. I don't have a strong preference, other than choosing one winner.

trentm commented on prototypes/prototype.adoc:
```diff
@@ -0,0 +1,18 @@
+    Copyright 2017 <contribution>
+////
+
+# RFD <Number> <Title>
```
Do we want "# ..." for the h1 or "= ..." as the AsciiDoc user-guide seems to suggest? (Note that 'make check' and the bot tooling don't yet recognize "= ...".)

sean- replied:
I'd advocate for `=` in order to be idiomatic.  The token alias is nice, but we shouldn't rely on it. 

> When in Rome, do as the Romans do.

sean- committed:
Update asciidoc prototype to include both the desired front-matter and attributes.

h/t to @jclulow for pointing the [`skip-front-matter`](http://asciidoctor.org/docs/user-manual/#front-matter-added-for-static-site-generators)
attribute.

sean- commented:
@jclulow Yup, I just got to that part of the manual.  Updated the `prototype.adoc` to match.

sean- committed:
Remove the `skip-front-matter` attribute, add a stub section.

The `skip-front-matter` attribute can only be passed in via the CLI:

```sh
$ asciidoctor --attribute=skip-front-matter ...
```

sean- committed:
Warn when an undefined attribute is referenced.

Brought to you by: future self asking, "WTF isn't this the default behavior?"

sean- committed:
Add keywords and a revnumber to the asciidoc prototype

sean- committed:
Fix admonition icons when rendering on Github

RFD's are going to need their own ghpages branch and
joyent.github.io/rfd/rfd*.html needs to be a thing otherwise this won't
render well by default.  This is a start, and it improves the state
of the world.

trentm commented on prototypes/prototype.adoc:
```diff
@@ -0,0 +1,44 @@
+---
+authors: Han Solo <han@example.com>, Frodo Baggins <frodo@example.com>
+state: predraft
+---
```
blank link before the title please... or perhaps let's move the title to after all the adoc attributes?

sean- replied:
@trentm np.  I'm curious as to why there is a preference.  I haven't read anything to suggest it matters to asciidoc.

trentm commented on prototypes/prototype.adoc:
```diff
@@ -0,0 +1,44 @@
+state: predraft
+---
+= RFD <NUMBER> <Title>
+:author: Han Solo <han@example.com>; Frodo Baggins <frodo@example.com>
```
Do we need this duplication? If there is a reason for 'author' and 'state' in adoc attributes, then I say we should drop the jekyll-format frontmatter.

sean- replied:
@trentm asciidoc does make use of this attribute somehow, somewhere.  If you look at the rendered HTML, it has a placeholder for this value:

<img width="951" alt="image" src="https://user-images.githubusercontent.com/349516/28794345-d7470506-75ea-11e7-9271-4209434a2d79.png">

The front-matter for jekyll is purely for Jekyll.  asciidoc ignores it completely (and though does save the value for reference, the content of the front-matter is un-indexable by asciidoc thereby rendering it effectively useless to asciidoc).  I'm indifferent to its presence despite the fact that it doesn't add any value to the meaning of the content and also doesn't improve Github's rendering of the document (compare GH's rendering of asciidoc vs the actual `asciidoctor`-rendered version at https://cdn.rawgit.com/joyent/rfd/1bc065db/rfd/0106/README.html).  I assumed the front-matter was being used somehow by one of our tools and was therefore a structurally important thing to leave in place for compatibility reasons.

trentm replied:
Yes, the front matter is being parsed by the chat bot, and I think by 'make check'. I think I'd prefer we just use the asciidoc attributes, and drop the jekyll front-matter for .adoc RFDs.

sean- replied:
I updated `rfdlint` to validate the titles for `.adoc` files.  `rfdlint` is using `=` as its doc title token, not `#`:  https://github.com/joyent/rfd/blob/master/tools/rfdlint#L131-L133

As for the chat bots, I haven't touched them but suspect it would be a similar change.

trentm replied:
Yup, thanks. I updated the new chatbot (joybot1999) already. I'll get the current jabber bot.

trentm replied:
I lied, I won't get the current jabber bot. Meh, if we are going to move away from it in the next couple of weeks.

trentm commented on prototypes/prototype.adoc:
```diff
@@ -0,0 +1,44 @@
+= RFD <NUMBER> <Title>
+:author: Han Solo <han@example.com>; Frodo Baggins <frodo@example.com>
+:state: predraft
+:revremark: State: {state}
```
I know/suspect this came from RFD 77's example. Do we care to have this? /cc @arekinath 
My goal here is to reduce boilerplate we don't need.

sean- replied:
`:revmark:` is the important attribute.  I think @arekinath's example was correct in teasing the `:state:` attribute out from the `:revmark:` attribute.

trentm commented on prototypes/prototype.adoc:
```diff
@@ -0,0 +1,44 @@
+:author: Han Solo <han@example.com>; Frodo Baggins <frodo@example.com>
+:state: predraft
+:revremark: State: {state}
+:attribute-missing: warn
```
Can we have a justification for any/all of these attributes? It would be nice if we could have *none* of them to have let distraction from the content when reading/editing the raw files.

sean- replied:
re: justification, yes.  Most of them substantially improve the readability of the document.  My understanding of this prototype was that it was intended to be the canonical source for a cargo-cult^Wskeleton RFD so that future authors wouldn't have to become experts and could still obtain the maximal benefit of the markup format (esp for larger/longer documents).

If we want to take a few minutes to optimize for readability of the raw files I'm sure we can figure out which attributes can be moved to the bottom of the document.  I have some anecdotal evidence from my reformatting effort that some of them can't be moved around arbitrarily.  I suspect - but have not confirmed - that `asciidoc` only makes a single-pass over the document.

In general I'd favor maintainability of raw files over readability of the raw files, especially in larger documents where consumption of the raw document is going to become increasingly difficult with the use of external files that are externally maintained and included at document compile-time.  I like that both Markdown and asciidoc are very readable forms in the raw (compared with `mdoc` or Docbook), but I'm less concerned with the first 50 lines of a file compared with the remaining hundreds to thousands of lines of the document that contain the real content.

sean- committed:
Use the `include` macro to move the asciidoc front-matter to an external file.

sean- commented:
@trentm and I had an offline conversation where he suggested `include`'ing the front-matter.  This appears to work nicely.

It still doesn't address the problem that Github doesn't render the includes inline (https://github.com/joyent/rfd/tree/master/rfd/0106#recommendations), but that's a secondary concern, imo, because now we have testable examples being used in documentation (https://github.com/joyent/rfd/blob/master/rfd/0106/examples/import_test.go).  The compiled HTML is still available via https://cdn.rawgit.com/joyent/rfd/2744c91/rfd/0106/README.html which is a marked improvement in nearly every dimension and its presentation is orthogonal to the prototype markup for asciidoc, imo.
